Copyright Jana Schaich Borg/Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)

# MySQL Exercise 8: Joining Tables with Outer Joins

Here's a picture to remind you of the general concepts behind outer joins:

<img src="https://duke.box.com/shared/static/4s89rmm8a75tep4puyqt1tdlhic0stzx.jpg" width=400 alt="SELECT FROM WHERE" />

To begin practicing outer joins, load the sql library, connect to the Dognition database, and make the Dognition database your default database: 


In [2]:
%load_ext sql
%sql mysql://studentuser:studentpw@mysqlserver/dognitiondb
%sql USE dognitiond

(_mysql_exceptions.OperationalError) (1044, "Access denied for user 'studentuser'@'%' to database 'dognitiond'") [SQL: 'USE dognitiond']


## Left and Right Joins

Left and right joins use a different sytax than we used in the lesson about inner joins.  The method I showed you to execute inner joins tells the database how to relate tables in a WHERE clause like this:

```mySQL
WHERE d.dog_guid=r.dog_guid
```

I find this syntax -- called the "equijoin" syntax -- to be very intuitive, so I thought it would be a good idea to start with it.  However, we can re-write the inner joins in the same syntax used by outer joins.  To use this more traditional syntax, you have to tell the database how to connect the tables using an ON clause that comes right after the FROM clause.  Make sure to specify the word "JOIN" explicitly.  This traditional version of the syntax frees up the WHERE clause for other things you might want to include in your query.  Here's what one of our queries from the inner join lesson would look like using the traditional syntax:

```mySQL
SELECT d.dog_guid AS DogID, d.user_guid AS UserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM dogs d JOIN reviews r
  ON d.dog_guid=r.dog_guid AND d.user_guid=r.user_guid
GROUP BY d.user_guid
HAVING NumRatings > 9
ORDER BY AvgRating DESC
LIMIT 200
```

You could also write "INNER JOIN" instead of "JOIN" but the default in MySQL is that JOIN will mean inner join, so including the word "INNER" is optional.

If you need a WHERE clause in the query above, it would go after the ON clause and before the GROUP BY clause.

Here's an example of a different query we used in the last lesson that employed the equijoin syntax:

```mySQL
SELECT d.user_guid AS UserID, d.dog_guid AS DogID, 
       d.breed, d.breed_type, d.breed_group
FROM dogs d, complete_tests c
WHERE d.dog_guid=c.dog_guid AND test_name='Yawn Warm-up';
```

**Question 1: How would you re-write this query using the traditional join syntax?**

In [3]:
%%sql
SELECT d.user_guid AS UserID, d.dog_guid AS DogID, d.breed, d.breed_type, d.breed_group
FROM dogs d JOIN complete_tests c  # the default JOIN is inner join
  ON d.dog_guid=c.dog_guid
WHERE test_name='Yawn Warm-up'
LIMIT 5;

5 rows affected.


[('ce134e42-7144-11e5-ba71-058fbc01cf0b', 'fd27b272-7144-11e5-ba71-058fbc01cf0b', 'Labrador Retriever', 'Pure Breed', 'Sporting'),
 ('ce1353d8-7144-11e5-ba71-058fbc01cf0b', 'fd27b5ba-7144-11e5-ba71-058fbc01cf0b', 'Shetland Sheepdog', 'Pure Breed', 'Herding'),
 ('ce135ab8-7144-11e5-ba71-058fbc01cf0b', 'fd27b6b4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'Pure Breed', 'Sporting'),
 ('ce13507c-7144-11e5-ba71-058fbc01cf0b', 'fd27b79a-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'Pure Breed', 'Sporting'),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Shih Tzu', 'Pure Breed', 'Toy')]

Let's start by using a left outer join to get the list we want. When we use the traditional join syntax to write inner joins, the order you enter the tables in your query doesn't matter. In outer joins, however, the order matters a lot. A left outer join will include all of the rows of the table to the left of the LEFT JOIN clause. A right outer join will include all of the rows of the table to the right of the RIGHT JOIN clause. So in order to retrieve a full list of dogs who completed at least 10 tests in the reviews table, and include as much breed information as possible, we could query:

```mysql
SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM reviews r LEFT JOIN dogs d
  ON r.dog_guid=d.dog_guid AND r.user_guid=d.user_guid
WHERE r.dog_guid IS NOT NULL
GROUP BY r.dog_guid
HAVING NumRatings >= 10
ORDER BY AvgRating DESC;
```

**Question 2: How could you retrieve this same information using a RIGHT JOIN?**


In [4]:
%%sql
SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM dogs d RIGHT JOIN reviews r # RIGHT JOIN IS associated with reviews table
  ON d.dog_guid=r.dog_guid AND d.user_guid=r.user_guid
WHERE r.dog_guid IS NOT NULL
GROUP BY r.dog_guid
HAVING NumRatings >= 10
ORDER BY AvgRating DESC
LIMIT 5;

5 rows affected.


[('fdbf39f8-7144-11e5-ba71-058fbc01cf0b', 'fdbf39f8-7144-11e5-ba71-058fbc01cf0b', 'ce987914-7144-11e5-ba71-058fbc01cf0b', 'ce987914-7144-11e5-ba71-058fbc01cf0b', Decimal('8.0000'), 12, 'Canaan Dog', 'Herding', 'Pure Breed'),
 ('ce47553e-7144-11e5-ba71-058fbc01cf0b', None, 'ce6ca9ba-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.8750'), 16, None, None, None),
 ('ce6f07e6-7144-11e5-ba71-058fbc01cf0b', None, 'ce7091e2-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.5000'), 10, None, None, None),
 ('ce45ae5a-7144-11e5-ba71-058fbc01cf0b', None, 'ce67562c-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.3529'), 17, None, None, None),
 ('ce2a68ac-7144-11e5-ba71-058fbc01cf0b', None, 'ce2a45c0-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.1333'), 15, None, None, None)]

Notice in the output of both the left and the right version of the outer join, all the rows that had a dog_guid in the reviews table but did NOT have a matching dog_guid in the dogs table have the word "None" entered in output columns related to the dogs table.  "None", in this case, is Jupyter's way of saying the value is NULL.  This becomes clear when you query a list of only the dog_guids that were NOT in the dogs table:

```mysql
SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM reviews r LEFT JOIN dogs d
  ON r.dog_guid=d.dog_guid AND r.user_guid=d.user_guid
WHERE d.dog_guid IS NULL
GROUP BY r.dog_guid
HAVING NumRatings >= 10
ORDER BY AvgRating DESC;
```

**Go ahead and try it yourself (you should get 894 rows in your query):**

In [5]:
%%sql
SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM reviews r LEFT JOIN dogs d
  ON r.dog_guid=d.dog_guid AND r.user_guid=d.user_guid
WHERE d.dog_guid IS NULL # all the rows that had a dog_guid in the reviews table but did NOT in the dogs table
GROUP BY r.dog_guid
HAVING NumRatings >= 10
ORDER BY AvgRating DESC
LIMIT 5;

5 rows affected.


[('ce47553e-7144-11e5-ba71-058fbc01cf0b', None, 'ce6ca9ba-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.8750'), 16, None, None, None),
 ('ce6f07e6-7144-11e5-ba71-058fbc01cf0b', None, 'ce7091e2-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.5000'), 10, None, None, None),
 ('ce45ae5a-7144-11e5-ba71-058fbc01cf0b', None, 'ce67562c-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.3529'), 17, None, None, None),
 ('ce2a68ac-7144-11e5-ba71-058fbc01cf0b', None, 'ce2a45c0-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.1333'), 15, None, None, None),
 ('ce72be0e-7144-11e5-ba71-058fbc01cf0b', None, 'ce73f7a6-7144-11e5-ba71-058fbc01cf0b', None, Decimal('7.0000'), 12, None, None, None)]

<img src="https://duke.box.com/shared/static/p2eucjdttai08eeo7davbpfgqi3zrew0.jpg" width=600 alt="SELECT FROM WHERE" />

**Question 3: How would you use a left join to retrieve a list of all the unique dogs in the dogs table, and retrieve a count of how many tests each one completed? Include the dog_guids and user_guids from the dogs and complete_tests tables in your output. (If you do not limit your query, your output should contain 35050 rows.  HINT: use the dog_guid from the dogs table to group your results.)**

In [6]:
%%sql
SELECT d.user_guid AS dUserID, c.user_guid AS cUserID, d.dog_guid AS dDogID, c.dog_guid AS cDogID, COUNT(test_name)
FROM dogs d LEFT JOIN complete_tests c
  ON d.dog_guid=c.dog_guid
GROUP BY dDogID
LIMIT 5;

5 rows affected.


[('ce134e42-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b272-7144-11e5-ba71-058fbc01cf0b', 'fd27b272-7144-11e5-ba71-058fbc01cf0b', 21),
 ('ce1353d8-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b5ba-7144-11e5-ba71-058fbc01cf0b', 'fd27b5ba-7144-11e5-ba71-058fbc01cf0b', 20),
 ('ce135ab8-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b6b4-7144-11e5-ba71-058fbc01cf0b', 'fd27b6b4-7144-11e5-ba71-058fbc01cf0b', 2),
 ('ce13507c-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b79a-7144-11e5-ba71-058fbc01cf0b', 'fd27b79a-7144-11e5-ba71-058fbc01cf0b', 11),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 31)]

Sometimes you can get so focused on writing your join statement that you don't pay close attention to the fields and tables you put in your other clauses, especially when you are joining a lot of tables.  Often your query will still run successfully, even if you haven't entered the criteria or grouping clause you intended.  The next question will illustrate how easy it is for this to happen.

**Question 4: Repeat the query you ran in Question 3, but intentionally use the dog_guids from the completed_tests table to group your results instead of the dog_guids from the dogs table. (Your output should contain 17987 rows)**

In [7]:
%%sql
SELECT d.user_guid AS dUserID, c.user_guid AS cUserID, d.dog_guid AS dDogID, c.dog_guid AS cDogID, COUNT(test_name)
FROM dogs d LEFT JOIN complete_tests c
  ON d.dog_guid=c.dog_guid
GROUP BY cDogID
LIMIT 5;

5 rows affected.


[('ce1362ba-7144-11e5-ba71-058fbc01cf0b', None, 'fd27baec-7144-11e5-ba71-058fbc01cf0b', None, 0),
 ('ce134e42-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b272-7144-11e5-ba71-058fbc01cf0b', 'fd27b272-7144-11e5-ba71-058fbc01cf0b', 21),
 ('ce1353d8-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b5ba-7144-11e5-ba71-058fbc01cf0b', 'fd27b5ba-7144-11e5-ba71-058fbc01cf0b', 20),
 ('ce135ab8-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b6b4-7144-11e5-ba71-058fbc01cf0b', 'fd27b6b4-7144-11e5-ba71-058fbc01cf0b', 2),
 ('ce13507c-7144-11e5-ba71-058fbc01cf0b', None, 'fd27b79a-7144-11e5-ba71-058fbc01cf0b', 'fd27b79a-7144-11e5-ba71-058fbc01cf0b', 11)]

This time your query ran successfully, but you retrieved many fewer DogIDs because the GROUP BY clause grouped your results according to the dog_guids in the completed_tests table rather than the dog_guid table.  As a result, even though you implemented your join correctly, all of the dog_guids that were in the dogs table but not in the completed_tests table got rolled up into one row of your output where completed_tests.dogs_guid = NULL.  This is a good opportunity to remind ourselves about the differences between SELECT/GROUP BY and COUNT DISTINCT.  

**Question 5: Write a query using COUNT DISTINCT to determine how many distinct dog_guids there are in the completed_tests table.**

In [8]:
%%sql
SELECT COUNT(DISTINCT dog_guid) # COUNT DISTINCT does NOT count NULL values, while SELECT/GROUP roll up NULL values into one group
FROM complete_tests;

1 rows affected.


[(17986,)]

**Question 6: We want to extract all of the breed information of every dog a user_guid in the users table owns.  If a user_guid in the users table does not own a dog, we want that information as well.  Write a query that would return this information.  Include the dog_guid from the dogs table, and user_guid from both the users and dogs tables in your output. (HINT: you should get 952557 rows in your output!)**  

In [9]:
%%sql
SELECT u.user_guid AS uUserID, d.user_guid AS dUserID, d.dog_guid AS dDogID, d.breed
FROM users u LEFT JOIN dogs d
  ON u.user_guid=d.user_guid
LIMIT 5;

5 rows affected.


[('ce134e42-7144-11e5-ba71-058fbc01cf0b', 'ce134e42-7144-11e5-ba71-058fbc01cf0b', 'fd27b272-7144-11e5-ba71-058fbc01cf0b', 'Labrador Retriever'),
 ('ce134e42-7144-11e5-ba71-058fbc01cf0b', 'ce134e42-7144-11e5-ba71-058fbc01cf0b', 'fd417cac-7144-11e5-ba71-058fbc01cf0b', 'Mixed'),
 ('ce1353d8-7144-11e5-ba71-058fbc01cf0b', 'ce1353d8-7144-11e5-ba71-058fbc01cf0b', 'fd27b5ba-7144-11e5-ba71-058fbc01cf0b', 'Shetland Sheepdog'),
 ('ce1353d8-7144-11e5-ba71-058fbc01cf0b', 'ce1353d8-7144-11e5-ba71-058fbc01cf0b', 'fd3fb0f2-7144-11e5-ba71-058fbc01cf0b', 'Shetland Sheepdog'),
 ('ce135ab8-7144-11e5-ba71-058fbc01cf0b', 'ce135ab8-7144-11e5-ba71-058fbc01cf0b', 'fd27b6b4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever')]

There are only 35050 distinct dog_guids in the dogs table.  Why is the database outputting almost a million rows?  That can't be right.  Let's figure out what is going on.  

**Question 7: Adapt the query you wrote above so that it counts the number of rows the join will output per user_id.  Sort the results by this count in descending order.  Remember that if you include dog_guid or breed fields in this query, they will be randomly populated by only one of the values associated with a user_guid (see MySQL Exercise 6; there should be 33,193 rows in your output).**

In [10]:
%%sql
SELECT u.user_guid AS uUserID, d.user_guid AS dUserID, d.dog_guid AS dDogID, d.breed, COUNT(*) AS numrows
FROM users u LEFT JOIN dogs d
  ON u.user_guid=d.user_guid
GROUP BY u.user_guid
ORDER BY numrows DESC
LIMIT 5;

5 rows affected.


[('ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b', 'fd7bfb52-7144-11e5-ba71-058fbc01cf0b', 'Shih Tzu', 913138),
 ('ce225842-7144-11e5-ba71-058fbc01cf0b', 'ce225842-7144-11e5-ba71-058fbc01cf0b', 'fd423714-7144-11e5-ba71-058fbc01cf0b', 'Shih Tzu', 442),
 ('ce2258a6-7144-11e5-ba71-058fbc01cf0b', 'ce2258a6-7144-11e5-ba71-058fbc01cf0b', 'fd40bd62-7144-11e5-ba71-058fbc01cf0b', 'Shih Tzu', 320),
 ('ce135e14-7144-11e5-ba71-058fbc01cf0b', 'ce135e14-7144-11e5-ba71-058fbc01cf0b', 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Shih Tzu', 130),
 ('ce29675e-7144-11e5-ba71-058fbc01cf0b', 'ce29675e-7144-11e5-ba71-058fbc01cf0b', 'fd46b014-7144-11e5-ba71-058fbc01cf0b', 'Labrador Retriever- Mix', 110)]

This query told us that user 'ce7b75bc-7144-11e5-ba71-058fbc01cf0b' would be associated with 913,138 rows in the output of the outer join we designed!  Once again, why?  We are going to work with the second user_guid in the output you just generated, 'ce225842-7144-11e5-ba71-058fbc01cf0b', because it would be associated with 442 output rows, and 442 rows are much easier to work with than 913,138.   

**Question 8: How many rows in the *users* table are associated with user_guid 'ce225842-7144-11e5-ba71-058fbc01cf0b'?**

In [11]:
%%sql
SELECT COUNT(*)
FROM users
WHERE user_guid='ce225842-7144-11e5-ba71-058fbc01cf0b';

1 rows affected.


[(17,)]

There are 17 entries associated with that user_guid in the users table.  If you examine all the columns in the entries, you will see that the rows are exact duplicates of each other.  That's unfortunate, but also something that can happen in real life data sets, especially those from new companies or governmental agencies.

Ok, now...

**Question 9: Examine all the rows in the *dogs* table that are associated with user_guid 'ce225842-7144-11e5-ba71-058fbc01cf0b'?**

In [12]:
%%sql
SELECT COUNT(*)
FROM dogs
WHERE user_guid='ce225842-7144-11e5-ba71-058fbc01cf0b';

1 rows affected.


[(26,)]

When databases join tables, they output the result of every pair of entries that meet certain criteria in the linking column of one table with the linking column of another table.  Our join statement imposed the criteria that the output should only include pairs whose user_guids matched in the two linking columns.  However, since there were multiple rows that had the same user_guid in the users table, *each one of these rows got paired up with each row in the dogs table that had the same user_guid.*  The result was 442 rows, because 17 (instances of the user_guid in the users table) x 26 (instances of the user_guid in the dogs table) = 442.

Outer joins are used very rarely.  The most practical application is if you want to export all of your raw data to another program for visualization or analysis. The syntax for outer joins is the same as for inner joins, but you replace the word "inner" with " full outer":

```mysql
SELECT r.dog_guid AS rDogID, d.dog_guid AS dDogID, r.user_guid AS rUserID, d.user_guid AS dUserID, AVG(r.rating) AS AvgRating, COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type
FROM reviews r FULL OUTER JOIN dogs d
  ON r.dog_guid=d.dog_guid AND r.user_guid=d.user_guid
WHERE r.dog_guid IS NOT NULL
GROUP BY r.dog_guid
ORDER BY AvgRating DESC;
```

<mark>**HOWEVER!  MySQL does not support full outer joins.**</mark>
   
If you wanted to imitate a full outer join in mySQL, you could follow one of the methods described at this website:

http://www.xaprb.com/blog/2006/05/26/how-to-write-full-outer-join-in-mysql/





## Practice outer joining your own tables!¶

**Question 10: How would you write a query that used a *left* join to return the number of distinct user_guids that were in the users table, but not the dogs table (your query should return a value of 2226)?**

In [17]:
%%sql
SELECT COUNT(DISTINCT u.user_guid) AS num_left
FROM users u LEFT JOIN dogs d
  ON u.user_guid=d.user_guid
WHERE d.user_guid IS NULL;

1 rows affected.


[(2226,)]

**Question 11: How would you write a query that used a *right* join to return the number of distinct user_guids that were in the users table, but not the dogs table (your query should return a value of 2226)?**

In [19]:
%%sql
SELECT COUNT(DISTINCT u.user_guid) AS num_right
FROM dogs d RIGHT JOIN users u
  ON d.user_guid=u.user_guid
WHERE d.user_guid IS NULL;

1 rows affected.


[(2226,)]

**Question 12: Use a left join to create a list of all the unique dog_guids that are contained in the site_activities table, but not the dogs table, and how many times each one is entered.  Note that there are a lot of NULL values in the dog_guid of the site_activities table, so you will want to exclude them from your list.  (Hint: if you exclude null values, the results you get will have two rows with words in their site_activities dog_guid fields instead of real guids, due to mistaken entries)**

In [20]:
%%sql
SELECT s.dog_guid AS unique_dogs_guid, COUNT(*) AS Num
FROM site_activities s LEFT JOIN dogs d
ON s.dog_guid=d.dog_guid
WHERE d.dog_guid IS NULL AND s.dog_guid IS NOT NULL
GROUP BY unique_dogs_guid;

2 rows affected.


[('Membership', 5587), ('PortalContent', 12)]